content/advent-2016/advanced-encoding-decoding.md

```diff
@@ -0,0 +1,507 @@
+
+```go
+type Dog struct {
+  ID     int       `json:"id"`
+  Name   string    `json:"name"`
+  Breed  string    `json:"breed"`
+  BornAt time.Time `json:"born_at"`
+}
+```
+
+We already know that this won't work, so rather than using the `time.Time` type, we are going to create our own `Time` type, embed the `time.Time` inside of the new type, and then update our `Dog` type to use our new `Time` type.
+
+```go
+type Dog struct {
+  ID     int    `json:"id"`
+  Name   string `json:"name"`
+  Breed  string `json:"breed"`
+  BornAt Time   `json:"born_at"`
+}
+
+type Time struct {
+  time.Time
+}
+```
+
+Next, we are going to write a custom `MarshalJSON()` and `UnmarshalJSON()` method for our `Time` type. These will output a unix time, and parse a unix time respectively.
+
+```go
+func (t Time) MarshalJSON() ([]byte, error) {
+  return json.Marshal(t.Time.Unix())
+}
+
+func (t *Time) UnmarshalJSON(data []byte) error {
+  var i int64
+  if err := json.Unmarshal(data, &i); err != nil {
+    return err
+  }
+  t.Time = time.Unix(i, 0)
+  return nil
+}
+```
+
+And that's it! We can now freely use our new `Time` type in all of our structs and it will be encoded and decoded as a unix time. On top of that, because we embedded the `time.Time` object, we can even freely use methods like `Day()` on our own `Time` type, meaning a good portion of our code won't need to be rewritten.
+
+That does bring up one of the downsides to this approach; By using a new type, we do stand the chance of breaking some of our code that expects the `time.Time` type rather than our new `Time` type. You can update all of your code to use the new type, or you can even access the embedded `time.Time` object, so it isn't impossible to fix this, but it may require a little bit of refactoring.
+
+Another solution to this problem is to merge this approach with the first one we looked at. That way we get the best of both worlds - our `Dog` type has a `time.Time` object, but our `JSONDog` doesn't need to worry itself with as many specifics for converting between the two types; The conversion logic is all contained within the new `Time` type.
+
+*A full example of this can be seen on the Go Playground here: <https://play.golang.org/p/C272eojwTh>*
+
+
+## Encoding and decoding generics
+
+The last technique we are going to look at is a little different than the first two we examined, and is meant to solve an entirely different problem - the problem of dynamic types being stored in nested JSON.
+
+For example, imagine you could get the following JSON responses from a server:
+
+```go
+{
+  "data": {
+    "object": "bank_account",
+    "id": "ba_123",
+    "routing_number": "110000000"
+  }
+}
+```
+
+And from the same endpoint you might also receive the following:
+
+```go
+{
+  "data": {
+    "object": "card",
+    "id": "card_123",
+    "last4": "4242"
+  }
+}
+```
```

Comment:
It's not super clear that these are different objects, it looks like they're the same object with an optional field of either `last4` or `routing_number`. It becomes evident later though. Perhaps just an extra sentence pointing out these are different objects with different fields, not the same object with optional fields. This is just my opinion and can be safely ignored.